apollo-runtime/src/jvmMain/kotlin/com/apollographql/apollo3/network/ws/OkHttpWebSocketEngine.kt

```diff
@@ -63,6 +64,13 @@ actual class DefaultWebSocketEngine(
         messageChannel.close(t)
       }
 
+      override fun onClosing(webSocket: WebSocket?, code: Int, reason: String?) {
+        webSocketOpenResult.complete(Unit)
+
+        val t = ApolloWebSocketException(code, reason)
+        messageChannel.close(t)
```

Comment:
Why does this need to be in the `onClosing` method ? I would expect that `onClosed` would be called ultimately and that should trigger a `ClosedReceiveChannelException` from `readWebSocket()`. Or is `onClosed()` never called in your case?

Reply:
In my testing, onClosing is only called when the server doesn't allow the connection e.g., expired token. I think it makes sense since that API includes a WebSocket code.

Reply:
Indeed. From the doc about `onClosing`:
```
Invoked when the remote peer has indicated that no more incoming messages will be transmitted.
```

So looks like `onClosed` = `onClosing` + client called `webSocket.close`? And the code in `onClosed` is the same as `onClosing`?

If that's the case, I think it should be fine since `messageChannel.close()` is idempotent. Being able to integration test this would be really nice. I created an issue there for this: https://github.com/apollographql/apollo-android/issues/3259